### PR TITLE
Storage APIs can throw arbitrary exceptions.  Be resilient to them.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Mitigation for https://github.com/dotnet/roslyn/issues/15780

**Customer scenario**

Customer upgrades from Dev14 to Dev15.  Loading any existing projects will crash.

**Bugs this fixes:** 

https://github.com/dotnet/roslyn/issues/15780

**Workarounds, if any**

Customer has to delete their .vs directory.  This is non discoverable, and sitll gives them a terrible upgrade expereince.

**Risk**
**Performance impact**

Low.  We're just being resilient to the lower level persistent layer failing to read/write data stored from a previous version of VS.

**Is this a regression from a previous update?**

Yes.  

**Root cause analysis:**

No testing of upgrade scenarios.  

**How was the bug found?**

Many Roslyn team members reported this upon syncing to a recent versin of Roslyn.  Due to a persistence format change, all the local caches stored from previous versions of VS were now causing VS to crash.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @MattGertz @heejaechang @mattwar

File: src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs [hunk 1]
```diff
@@ -22,6 +24,121 @@ internal sealed partial class SyntaxTreeIndex : AbstractPersistableState, IObjec
...
+        }
+
+        private static async Task<SyntaxTreeIndex> LoadAsync(
+            Document document, string persistenceName, string formatVersion,
+            Func<ObjectReader, VersionStamp, SyntaxTreeIndex> readFrom, CancellationToken cancellationToken)
+        {
+            var persistentStorageService = document.Project.Solution.Workspace.Services.GetService<IPersistentStorageService>();
+            var syntaxVersion = await document.GetSyntaxVersionAsync(cancellationToken).ConfigureAwait(false);
+
+            try
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs]: This is a new try/catch.

File: src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs [hunk 2]
```diff
@@ -22,6 +24,121 @@ internal sealed partial class SyntaxTreeIndex : AbstractPersistableState, IObjec
...
+        }
+
+        private static async Task<bool> SaveAsync(
+            Document document, string persistenceName, string formatVersion, SyntaxTreeIndex data, CancellationToken cancellationToken)
+        {
+            Contract.Requires(!await document.IsForkedDocumentWithSyntaxChangesAsync(cancellationToken).ConfigureAwait(false));
+
+            var persistentStorageService = document.Project.Solution.Workspace.Services.GetService<IPersistentStorageService>();
+
+            try
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs]: This is a new try/catch.

File: src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs [hunk 3]
```diff
@@ -22,6 +24,121 @@ internal sealed partial class SyntaxTreeIndex : AbstractPersistableState, IObjec
...
+
+        private static async Task<bool> PrecalculatedAsync(Document document, string persistenceName, string formatVersion, CancellationToken cancellationToken)
+        {
+            Contract.Requires(document.IsFromPrimaryBranch());
+
+            var persistentStorageService = document.Project.Solution.Workspace.Services.GetService<IPersistentStorageService>();
+            var syntaxVersion = await document.GetSyntaxVersionAsync(cancellationToken).ConfigureAwait(false);
+
+            // check whether we already have info for this document
+            try
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs]: This is a new try/catch.

**Comment by MattGertz** [comment]
(Please retag me once CRs are done.)

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide Need PR.  This is for a crash that will affect everyone (including you) if you roll forward to a new version of Roslyn.

File: src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs [hunk 4]
```diff
@@ -22,6 +24,121 @@ internal sealed partial class SyntaxTreeIndex : AbstractPersistableState, IObjec
...
+                            document.CanReusePersistedSyntaxTreeVersion(syntaxVersion, persistVersion))
+                        {
+                            return readFrom(reader, syntaxVersion);
+                        }
+                    }
+                }
+            }
+            catch (Exception)
+            {
+                // Storage APIs can throw arbitrary exceptions.
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs]: Should we say something like "if this isn't an IO exception", non-fatal? I'm guessing you've already done this thought experiment, but a comment here might be useful.

**Comment by heejaechang** [comment]
👍 

as stream format changes, object reader from previous format could cause random exception to happen. that is the underlying issue.

see this #15780

**Comment by CyrusNajmabadi** [comment]
> Actually on second thought, dismissingh approval for a second. Why was the underlying code throwing? Because of IO or because something else wasn't robust?

StreamObjectReader/Writer make no guarantees they will work on streams created by previous versions of this API.  The API does no validation, and stores no information to let it detect that it's on the wrong version.  @mattwar wants to keep it this way for performance reasons.   I don't personally agree, and i think the stream can at least have a starting version number that can be reasonably checked.

however, in the meantime, we need something to prevent the crashing that will happen.  We're catching all exceptions simply because StreamObjectReader does not document the exceptions it may throw.  So there's no way for us to limit the set unless that list is actually documented. :-/

**Comment by mattwar** [comment]
:+1:

**Comment by CyrusNajmabadi** [comment]
Do you have an alternate suggestion?  Right now we have an API that throws exceptions that are not documented.

**Comment by tmat** [comment]
What API is that?

**Comment by CyrusNajmabadi** [comment]
>What API is that?

StreamObjectReader.  Currently it throws when it changes its internal data format.  But there are no APIs provided to detect that.  And it does not document what exceptions it may throw.

**Comment by tmat** [comment]
I'm confused. StreamObjectReader is defined in Workspace layer.

File: src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs]: It's not IO exceptions.  From perusing the code, i can see at least the following:
InvalidOperationException
ArgumentException
RecursionDepthExceeded
UnreachableException

**Comment by CyrusNajmabadi** [comment]
> I'm confused. StreamObjectReader is defined in Workspace layer.

And...?  It does not document which exceptions it may throw.  It provides no entrypoint to attempt to use it in a non-exception-throwing manner.  How could we write code that properly uses it when that set of exceptions coudl change at any point in time (i.e. like it did just now)?

I have asked for improvements to that API.  But *in the meantime*, we need to addrss this or else anyone dogfooding on the team will face crashes.

**Comment by CyrusNajmabadi** [comment]
If someone is willing to pick up fixing the underlying API issue, i'm happy to close this PR.  But if that can't be done *tonight* then i really want this to go in to prevent more pain on the team.

**Comment by CyrusNajmabadi** [comment]
Note: the first line of the PR: Mitigation for #15780

This is not a fix.  This is a mitigation for an unfortunate situation.  It's the simplest and cheapest mitigation presently available until we come up with a better story for StreamObjectReader/Writer such that it is properly documented and can be used by the rest of hte stack safely.

**Comment by tmat** [comment]
I see. File a new bug to clean this up?

**Comment by CyrusNajmabadi** [comment]
We have the bug :)  It is #15780

**Comment by CyrusNajmabadi** [comment]
@MattGertz  CRs are done.

**Comment by tmat** [comment]
@CyrusNajmabadi  OK, just don't close it when merging this PR :)

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
retest windows_release_unit32_prtest please

**Pull request merged**